hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -993,6 +993,244 @@ class Emit[C](
+              (x.pType.asInstanceOf[PTuple].types(1).asInstanceOf[PNDArray], K, N)
+            } else {
+              throw new AssertionError(s"Unsupported QR mode $mode")
+            }
+
+            val rShapeArray = FastIndexedSeq[Value[Long]](rRows, rCols)
+
+            val rShapeBuilder = rPType.makeShapeBuilder(rShapeArray)
+            val rStridesBuilder = rPType.makeColumnMajorStridesBuilder(rShapeArray, mb)
+
+            // This block assumes that `rDataAddress` and `aAddressDGEQRF` point to column major arrays.
+            // TODO: Abstract this into ndarray ptype/pcode interface methods.
+            val copyOutUpperTriangle =
+            Code.forLoop(currCol := 0, currCol < rCols.toI, currCol := currCol + 1,
+              Code.forLoop(currRow := 0, currRow < rRows.toI, currRow := currRow + 1,
+                Region.storeDouble(
+                  pndValue.pt.data.pType.elementOffset(rDataAddress, aNumElements.toI, currCol * rRows.toI + currRow),
+                  (currCol >= currRow).mux(
+                    Region.loadDouble(pndValue.pt.data.pType.elementOffset(aAddressDGEQRF, aNumElements.toI, currCol * M.toI + currRow)),
+                    0.0
+                  )
+                )
+              )
+            )
+            val computeR = Code(
+              rDataAddress := rPType.data.pType.allocate(region.code, aNumElements.toI),
+              rPType.data.pType.stagedInitialize(rDataAddress, (rRows * rCols).toI),
+              copyOutUpperTriangle,
+              rPType.construct(rShapeBuilder, rStridesBuilder, rDataAddress, mb, region.code)
+            )
+
+            if (mode == "r") {
+              Code(
+                computeHAndTau,
+                computeR
+              )
+            }
+            else {
+              val crPType = x.pType.asInstanceOf[PTuple]
+              val crOutputSrvb = new StagedRegionValueBuilder(mb, crPType, region.code)
+
+              val qPType = crPType.types(0).asInstanceOf[PNDArray]
+              val qShapeArray = if (mode == "complete") Array(M, M) else Array(M, K)
+              val qShapeBuilder = qPType.makeShapeBuilder(qShapeArray)
+              val qStridesBuilder = qPType.makeColumnMajorStridesBuilder(qShapeArray, mb)
+
+              val rNDArrayAddress = mb.genFieldThisRef[Long]()
+              val qDataAddress = mb.genFieldThisRef[Long]()
+
+              val infoDORGQRResult = mb.genFieldThisRef[Int]()
+              val infoDORQRErrorTest = (extraErrorMsg: String) => (infoDORGQRResult cne 0)
+                .orEmpty(Code._fatal[Unit](const(s"LAPACK error DORGQR. $extraErrorMsg Error code = ").concat(infoDORGQRResult.toS)))
+
+              val qCondition = mb.genFieldThisRef[Boolean]()
+              val numColsToUse = mb.genFieldThisRef[Long]()
+              val aAddressDORGQR = mb.genFieldThisRef[Long]()
+
+              val qNumElements = mb.genFieldThisRef[Long]()
+
+              val computeCompleteOrReduced = Code(Code(FastIndexedSeq(
```

Comment:
This is another long stretch of code that could largely be placed on the CodeBuilder.